internal/executor/instance/persistentworker/isolation/parallels/parallels.go

```diff
@@ -73,122 +67,9 @@ func (parallels *Parallels) Run(ctx context.Context, config *runconfig.RunConfig
-		},
-	}
-
-	sshConn, chans, reqs, err := ssh.NewClientConn(netConn, addr, sshConfig)
-	if err != nil {
-		return fmt.Errorf("%w: failed to connect to the VM %q via SSH: %v", ErrFailed, vm.Ident(), err)
-	}
-
-	cli := ssh.NewClient(sshConn, chans, reqs)
-
-	// Work around x/crypto/ssh not being context.Context-friendly (e.g. https://github.com/golang/go/issues/20288)
-	monitorCtx, monitorCancel := context.WithCancel(ctx)
-	go func() {
-		<-monitorCtx.Done()
-		_ = cli.Close()
-	}()
-	defer monitorCancel()
-
-	remoteAgentPath, err := uploadAgent(ctx, cli, parallels.agentOS, config.GetAgentVersion())
-	if err != nil {
-		return fmt.Errorf("%w: failed to upload agent to the VM %q via SFTP: %v",
-			ErrFailed, vm.Ident(), err)
-	}
-
-	sess, err := cli.NewSession()
-	if err != nil {
-		return fmt.Errorf("%w: failed to open SSH session on VM %q: %v", ErrFailed, vm.Ident(), err)
-	}
-
-	// Log output from the virtual machine
-	stdout, err := sess.StdoutPipe()
-	if err != nil {
-		return fmt.Errorf("%w: while opening stdout pipe: %v", ErrFailed, err)
-	}
-	stderr, err := sess.StderrPipe()
-	if err != nil {
-		return fmt.Errorf("%w: while opening stderr pipe: %v", ErrFailed, err)
-	}
-	go func() {
-		output := io.MultiReader(stdout, stderr)
-
-		scanner := bufio.NewScanner(output)
-
-		for scanner.Scan() {
-			parallels.logger.Debugf("VM: %s", scanner.Text())
-		}
-	}()
-
-	stdinBuf, err := sess.StdinPipe()
-	if err != nil {
-		return fmt.Errorf("%w: while opening stdin pipe: %v", ErrFailed, err)
-	}
-
-	// start a login shell so all the customization from ~/.zprofile will be picked up
-	err = sess.Shell()
-	if err != nil {
-		return fmt.Errorf("%w: failed to start a shell on VM %q: %v", ErrFailed, vm.Ident(), err)
-	}
-
-	// Synchronize time for suspended VMs
-	if vm.ClonedFromSuspended() {
-		_, err = stdinBuf.Write([]byte(TimeSyncCommand(time.Now().UTC())))
-		if err != nil {
-			return fmt.Errorf("%w: failed to sync time on VM %q: %v", ErrFailed, vm.Ident(), err)
-		}
-	}
-
-	command := []string{
-		remoteAgentPath,
-		"-api-endpoint",
-		config.DirectEndpoint,
-		"-server-token",
-		"\"" + config.ServerSecret + "\"",
-		"-client-token",
-		"\"" + config.ClientSecret + "\"",
-		"-task-id",
-		strconv.FormatInt(config.TaskID, 10),
-	}
-
-	// Start the agent and wait for it to terminate
-	_, err = stdinBuf.Write([]byte(strings.Join(command, " ") + "\nexit\n"))
-	if err != nil {
-		return fmt.Errorf("%w: failed to start agent on VM %q: %v", ErrFailed, vm.Ident(), err)
-	}
-	err = sess.Wait()
-	if err != nil {
-		// Work around x/crypto/ssh not being context.Context-friendly (e.g. https://github.com/golang/go/issues/20288)
-		if err := monitorCtx.Err(); err != nil {
-			return err
-		}
-
-		return fmt.Errorf("%w: failed to run agent on VM %q: %v", ErrFailed, vm.Ident(), err)
-	}
-
-	return nil
+	return remoteagent.WaitForAgent(ctx, parallels.logger, ip,
+		parallels.sshUser, parallels.sshPassword, parallels.agentOS, "amd64",
```

Comment:
I think it's safe to pass `runtime.GOARCH` instead of `"amd64"` here. Parallels on M1 only can ran Arm VMs.

Reply:
>Parallels on M1 only can ran Arm VMs.

I think that's exactly the reason why it should be pegged to a specific architecture instead of being dynamic.

We explicitly signal to the reader that it only works on `amd64`.

Reply:
Oh, I've re-read the sentence and assuming that we're going to support Parallels on M1 it makes sense to switch `runtime.GOARCH` here.

But are we going to support it? Otherwise we should instead place a guard somewhere here to prevent running on non-`amd64`: https://github.com/cirruslabs/cirrus-cli/blob/c1b4602d37df924a3177b9c8272c1424157db02f/internal/executor/instance/persistentworker/persistentworker.go#L27-L29

Reply:
Sounds good. Let's error that it's not supported and please use tart.